intermediate_source/transformer_building_blocks.py

```diff
@@ -0,0 +1,641 @@
+# SwiGLU is a non-linear activation function that is increasingly popular in the feed-forward
+# network of the transformer layer (e.g. Llama). A FFN with SwiGLU activation is defined as
+
+class SwiGLUFFN(nn.Module):
+    def __init__(self, dim, hidden_dim, multiple_of, ffn_dim_multiplier=None, device=None, dtype=None):
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super().__init__()
+        hidden_dim = int(2 * hidden_dim / 3)
+        # custom dim factor multiplier
+        if ffn_dim_multiplier is not None:
+            hidden_dim = int(ffn_dim_multiplier * hidden_dim)
+        hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) // multiple_of)
+
+        self.w1 = nn.Linear(dim, hidden_dim, bias=False, **factory_kwargs)
+        self.w2 = nn.Linear(hidden_dim, dim, bias=False, **factory_kwargs)
+        self.w3 = nn.Linear(dim, hidden_dim, bias=False, **factory_kwargs)
+    
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+# An alternative way of implementing this that uses packed projection is 
+
+class PackedSwiGLUFFN(nn.Module):
+    def __init__(self, dim, hidden_dim, multiple_of, ffn_dim_multiplier=None, device=None, dtype=None):
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super().__init__()
+        hidden_dim = int(2 * hidden_dim / 3)
+        # custom dim factor multiplier
+        if ffn_dim_multiplier is not None:
+            hidden_dim = int(ffn_dim_multiplier * hidden_dim)
+        hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) // multiple_of)
+
+        self.w13 = nn.Linear(dim, 2 * hidden_dim, bias=False, **factory_kwargs)
+        self.w2 = nn.Linear(hidden_dim, dim, bias=False, **factory_kwargs)
+    
+    def forward(self, x):
+        x1, x3 = torch.chunk(self.w13(x), 2, dim=-1)
+        return self.w2(F.silu(x1) * x3)
+
+# We can compare the performance of the two implementations as follows
+# Depending on your hardware, you might see different results. On an A100 I see
+# 1.12x speedup for D=128.
+D = 128
+
+swigluffn = torch.compile(SwiGLUFFN(D, D * 4, 256, device='cuda', dtype=torch.bfloat16))
+packed_swigluffn = torch.compile(PackedSwiGLUFFN(D, D * 4, 256, device='cuda', dtype=torch.bfloat16))
+
+q, _, _, sentence_lengths = gen_batch(D, D, D, D, device="cuda", dtype=torch.bfloat16)
+
+# warmup
+swigluffn(q)
+packed_swigluffn(q)
+
+# benchmark
+_, time, _ = benchmark(swigluffn, q)
+_, time_packed, _ = benchmark(packed_swigluffn, q)
+print(f"SwiGLUFFN: {time} s, PackedSwiGLUFFN: {time_packed} s, speedup: {time/time_packed:.2f}x")
+
+###############################################################################
+# Cross Attention
+# ---------------
+# Cross attention is a form of attention where the query and key/value tensors
+# are from different sequences.
+
+# One example of this is in ``nn.TransformerDecoderLayer`` where the query comes
+# from the decoder and the key/value come from the encoder.
```

Comment:
Does this actually work with NJTs? lol I'm surprised you're not running into problems with having different nested ints for query / key+value

Reply:
Hm, running this code doesn't seem to error for me

Also in TransformerDecoderLayer's mha block we see this pattern as well, but I don't think I ran into errors with it when sanity check testing https://github.com/mikaylagawarecki/temp/blob/main/sanity_check_test.py#L241

Reply:
Resolving: From offline discussion, this is sanity checked in the above link. We were also concerned that the sdpa MATH path might not be operating correctly for this snippet but double checked that as well

```python
torch.manual_seed(6)
N, E_q, E_k, E_v, E_total = 512, 512, 512, 512, 512
device='cuda'
nheads = 8
E_head = 512 // 8
bias = True
dropout = 0.0
mha_layer = MultiHeadAttention(E_q, E_k, E_v, E_total, nheads, dropout=dropout, bias=bias, device='cuda')

mha_layer = torch.compile(mha_layer)

query, _, _, q_len = gen_batch(N, E_q, E_k, E_v, device)
_, key, value, kv_len = gen_batch(N, E_q, E_k, E_v, device)


# Total sequence length in nested query 11341, max sequence length 139
# Total sequence length in nested key/value 11446, max sequence length 146
print(f"Total sequence length in nested query {q_len.sum().item()}, max sequence length {q_len.max().item()}")
print(f"Total sequence length in nested key/value {kv_len.sum().item()}, max sequence length {kv_len.max().item()}")

with sdpa_kernel([SDPBackend.EFFICIENT_ATTENTION]):
    out_efficient = mha_layer(query, key, value, is_causal=False)

with sdpa_kernel([SDPBackend.MATH]):
    out_math = mha_layer(query, key, value, is_causal=False)

# 4.76837158203125e-07
print((out_efficient.to_padded_tensor(0.0) - out_math.to_padded_tensor(0.0)).max().item())
```